File: frontend/lib/core/services/room_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../constants/app_config.dart';
import '../storage/token_storage.dart';

class RoomServiceException implements Exception {
  RoomServiceException(this.message);

  final String message;

  @override
  String toString() => 'RoomServiceException: $message';
}

class CreatedRoom {
  const CreatedRoom({
    required this.id,
    required this.name,
    required this.roomType,
    this.description,
    this.avatarUrl,
    this.ownerId,
  });

  final String id;
  final String name;
  final String roomType;
  final String? description;
  final String? avatarUrl;
  final String? ownerId;

  factory CreatedRoom.fromJson(Map<String, dynamic> json) {
    return CreatedRoom(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      roomType: json['room_type'] as String? ?? 'group',
      description: json['description'] as String?,
      avatarUrl: json['avatar_url'] as String?,
      ownerId: json['owner_id'] as String?,
    );
  }
}

class GroupSettingsInfo {
  GroupSettingsInfo({
    required this.roomId,
    required this.globalMuteEnabled,
    this.globalMuteReason,
    this.globalMuteUntil,
    this.joinApprovalRequired,
    this.memberCanInvite,
    this.maxMembers,
  });

  final String roomId;
  final bool globalMuteEnabled;
  final String? globalMuteReason;
  final DateTime? globalMuteUntil;
  final bool? joinApprovalRequired;
  final bool? memberCanInvite;
  final int? maxMembers;

  factory GroupSettingsInfo.fromJson(Map<String, dynamic> json) {
    DateTime? muteUntil;
    final until = json['global_mute_until'];
    if (until is String && until.isNotEmpty) {
      muteUntil = DateTime.tryParse(until);
    }

    return GroupSettingsInfo(
      roomId: json['room_id'] as String? ?? '',
      globalMuteEnabled: json['global_mute_enabled'] as bool? ?? false,
      globalMuteReason: json['global_mute_reason'] as String?,
      globalMuteUntil: muteUntil,
      joinApprovalRequired: json['join_approval_required'] as bool?,
      memberCanInvite: json['member_can_invite'] as bool?,
      maxMembers: json['max_members'] as int?,
    );
  }
}

class RoomService {
  RoomService({TokenStorage? tokenStorage})
    : _tokenStorage = tokenStorage ?? const TokenStorage();

  final TokenStorage _tokenStorage;

  Future<Map<String, String>> _authHeaders() async {
    final session = await _tokenStorage.readSession();
    if (session == null) {
      throw RoomServiceException('用户未登录');
    }

    return {
      'Authorization': 'Bearer ${session.token}',
      'Content-Type': 'application/json',
    };
  }

  Future<CreatedRoom> createGroup({
    required String name,
    String? description,
    List<String> memberIds = const [],
  }) async {
    if (name.trim().isEmpty) {
      throw RoomServiceException('群聊名称不能为空');
    }
    if (memberIds.isEmpty) {
      throw RoomServiceException('请至少选择一位好友');
    }

    final headers = await _authHeaders();
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/rooms');
    final payload = <String, dynamic>{
      'name': name.trim(),
      'room_type': 'group',
      'member_ids': memberIds,
    };
    if (description != null && description.trim().isNotEmpty) {
      payload['description'] = description.trim();
    }

    final response = await http.post(
      uri,
      headers: headers,
      body: jsonEncode(payload),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      if (data is Map<String, dynamic>) {
        final roomData = data['room'];
        if (roomData is Map<String, dynamic>) {
          return CreatedRoom.fromJson(roomData);
        }
      }
      throw RoomServiceException('创建群聊返回数据格式异常');
    }

    throw RoomServiceException(_extractErrorMessage(response.body) ?? '创建群聊失败');
  }

  String? _extractErrorMessage(String body) {
    try {
      final data = jsonDecode(body);
      if (data is Map<String, dynamic>) {
        final message = data['message'];
        if (message is String && message.isNotEmpty) {
          return message;
        }
        final error = data['error'];
        if (error is String && error.isNotEmpty) {
          return error;
        }
      }
    } catch (_) {
      return null;
    }
    return null;
  }

  Future<GroupSettingsInfo> fetchGroupSettings(String roomId) async {
    if (roomId.isEmpty) {
      throw RoomServiceException('无效的群组 ID');
    }
    final headers = await _authHeaders();
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/rooms/$roomId/settings');
    final response = await http.get(uri, headers: headers);

    if (response.statusCode == 200) {
      final decoded = jsonDecode(response.body);
      if (decoded is Map<String, dynamic>) {
        final settings = decoded['settings'];
        if (settings is Map<String, dynamic>) {
          return GroupSettingsInfo.fromJson(settings);
        }
      }
      throw RoomServiceException('群设置返回数据异常');
    }

    throw RoomServiceException(
      _extractErrorMessage(response.body) ?? '加载群设置失败',
    );
  }

  Future<void> updateGlobalMute({
    required String roomId,
    required bool enabled,
    String? reason,
    int? durationMinutes,
  }) async {
    if (roomId.isEmpty) {
      throw RoomServiceException('无效的群组 ID');
    }
    final headers = await _authHeaders();
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/rooms/$roomId/mutes/global');
    final payload = <String, dynamic>{'enabled': enabled};
    if (reason != null && reason.trim().isNotEmpty) {
      payload['reason'] = reason.trim();
    }
    if (durationMinutes != null) {
      payload['duration_minutes'] = durationMinutes;
    }

    final response = await http.post(
      uri,
      headers: headers,
      body: jsonEncode(payload),
    );

    if (response.statusCode != 200) {
      throw RoomServiceException(
        _extractErrorMessage(response.body) ?? '更新群禁言状态失败',
      );
    }
  }

  Future<void> dissolveGroup(String roomId) async {
    if (roomId.isEmpty) {
      throw RoomServiceException('无效的群组 ID');
    }
    final headers = await _authHeaders();
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/rooms/$roomId');
    final response = await http.delete(uri, headers: headers);

    if (response.statusCode != 200) {
      throw RoomServiceException(
        _extractErrorMessage(response.body) ?? '解散群聊失败',
      );
    }
  }

  Future<void> transferRoomOwner({
    required String roomId,
    required String newOwnerId,
  }) async {
    if (roomId.isEmpty || newOwnerId.trim().isEmpty) {
      throw RoomServiceException('请先选择新的群主');
    }
    final headers = await _authHeaders();
    final uri = Uri.parse('${AppConfig.apiBaseUrl}/rooms/$roomId/transfer');
    final response = await http.post(
      uri,
      headers: headers,
      body: jsonEncode({'new_owner_id': newOwnerId.trim()}),
    );

    if (response.statusCode != 200) {
      throw RoomServiceException(
        _extractErrorMessage(response.body) ?? '转让群主失败',
      );
    }
  }
}

File: h5-app/src/services/room-service.ts
import { requestJson, withQuery } from '@/api/http';
import type { AddMembersResult, CreatedRoom, GroupAdmin, GroupDirectoryEntry, GroupInvitation, GroupInvitationStatus, GroupJoinRequest, GroupMute, GroupOperationLog, GroupRule, GroupSettingsInfo, JoinRequestStatus, RoomMember } from '@/types/room';

import { mapAddMembersResult, mapCreatedRoom, mapGroupAdmin, mapGroupInvitation, mapGroupJoinRequest, mapGroupMute, mapGroupOperationLog, mapGroupRule, mapGroupSettings, mapRoomMember } from './mappers';
import { requireToken } from './session';

export const roomService = {
  async createGroup(params: { name: string; description?: string; memberIds: string[] }): Promise<CreatedRoom> {
    const response = await requestJson<{ room?: Record<string, unknown> } & Record<string, unknown>>('/rooms', {
      method: 'POST',
      body: JSON.stringify({
        name: params.name.trim(),
        room_type: 'group',
        member_ids: params.memberIds,
        ...(params.description?.trim() ? { description: params.description.trim() } : {}),
      }),
    }, requireToken());
    return mapCreatedRoom(response.room ?? response);
  },

  async listRooms(): Promise<CreatedRoom[]> {
    const response = await requestJson<Record<string, unknown>[] | { rooms?: Record<string, unknown>[] }>(
      '/rooms',
      {},
      requireToken(),
    );
    const rooms = Array.isArray(response) ? response : response.rooms ?? [];
    return rooms.map(mapCreatedRoom);
  },

  async listGroupDirectory(): Promise<GroupDirectoryEntry[]> {
    const rows = await requestJson<Record<string, unknown>[]>('/groups/directory', {}, requireToken());
    return rows.map((row) => ({
      roomId: String(row.room_id ?? ''), name: String(row.name ?? ''),
      description: row.description == null ? null : String(row.description),
      avatarUrl: row.avatar_url == null ? null : String(row.avatar_url),
      avatarObjectKey: row.avatar_object_key == null ? null : String(row.avatar_object_key),
      memberCount: Number(row.member_count ?? 0), isFavorited: Boolean(row.is_favorited),
      favoritedAt: row.favorited_at == null ? null : String(row.favorited_at),
    }));
  },

  async favoriteGroupDirectory(roomId: string, favorited: boolean): Promise<void> {
    await requestJson(`/rooms/${roomId}/directory-favorite`, { method: favorited ? 'POST' : 'DELETE' }, requireToken());
  },

  async getRoom(roomId: string): Promise<CreatedRoom> {
    const response = await requestJson<{ room?: Record<string, unknown> } & Record<string, unknown>>(
      `/rooms/${roomId}`,
      {},
      requireToken(),
    );
    return mapCreatedRoom(response.room ?? response);
  },

  async updateRoom(roomId: string, params: { name?: string; description?: string }): Promise<CreatedRoom> {
    const response = await requestJson<{ room?: Record<string, unknown> } & Record<string, unknown>>(`/rooms/${roomId}`, {
      method: 'PATCH',
      body: JSON.stringify({
        ...(params.name !== undefined ? { name: params.name.trim() } : {}),
        ...(params.description !== undefined ? { description: params.description.trim() } : {}),
      }),
    }, requireToken());
    return mapCreatedRoom(response.room ?? response);
  },

  async listMembers(roomId: string): Promise<RoomMember[]> {
    const response = await requestJson<Record<string, unknown>[] | { members?: Record<string, unknown>[] }>(
      `/rooms/${roomId}/members`,
      {},
      requireToken(),
    );
    const members = Array.isArray(response) ? response : response.members ?? [];
    return members.map(mapRoomMember);
  },

  async addMembers(roomId: string, userIds: string[]): Promise<AddMembersResult> {
    const response = await requestJson<Record<string, unknown>>(`/rooms/${roomId}/members/add`, {
      method: 'POST',
      body: JSON.stringify({ user_ids: userIds }),
    }, requireToken());
    return mapAddMembersResult(response);
  },

  async removeMember(roomId: string, userId: string): Promise<void> {
    await requestJson(`/rooms/${roomId}/members/${userId}`, { method: 'DELETE' }, requireToken());
  },

  async listAdmins(roomId: string): Promise<GroupAdmin[]> {
    const response = await requestJson<{ admins?: Record<string, unknown>[] }>(
      `/rooms/${roomId}/admins`,
      {},
      requireToken(),
    );
    return (response.admins ?? []).map(mapGroupAdmin);
  },

  async appointAdmin(roomId: string, userId: string): Promise<GroupAdmin> {
    const response = await requestJson<{ admin: Record<string, unknown> }>(`/rooms/${roomId}/admins`, {
      method: 'POST',
      body: JSON.stringify({ user_id: userId, role: 'admin' }),
    }, requireToken());
    return mapGroupAdmin(response.admin);
  },

  async removeAdmin(roomId: string, userId: string): Promise<void> {
    await requestJson(`/rooms/${roomId}/admins/${userId}`, { method: 'DELETE' }, requireToken());
  },

  async listRules(roomId: string): Promise<GroupRule[]> {
    const response = await requestJson<{ rules?: Record<string, unknown>[] }>(
      `/rooms/${roomId}/rules`,
      {},
      requireToken(),
    );
    return (response.rules ?? []).map(mapGroupRule);
  },

  async createRule(roomId: string, input: { title: string; content: string; orderIndex: number }): Promise<GroupRule> {
    const response = await requestJson<{ rule: Record<string, unknown> }>(`/rooms/${roomId}/rules`, {
      method: 'POST',
      body: JSON.stringify({
        title: input.title.trim(),
        content: input.content.trim(),
        order_index: input.orderIndex,
      }),
    }, requireToken());
    return mapGroupRule(response.rule);
  },

  async updateRule(roomId: string, ruleId: string, input: { title: string; content: string }): Promise<GroupRule> {
    const response = await requestJson<{ rule: Record<string, unknown> }>(`/rooms/${roomId}/rules/${ruleId}`, {
      method: 'PATCH',
      body: JSON.stringify({ title: input.title.trim(), content: input.content.trim() }),
    }, requireToken());
    return mapGroupRule(response.rule);
  },

  async deleteRule(roomId: string, ruleId: string): Promise<void> {
    await requestJson(`/rooms/${roomId}/rules/${ruleId}`, { method: 'DELETE' }, requireToken());
  },

  async listMutes(roomId: string): Promise<GroupMute[]> {
    const response = await requestJson<{ mutes?: Record<string, unknown>[] }>(
      `/rooms/${roomId}/mutes`,
      {},
      requireToken(),
    );
    return (response.mutes ?? []).map(mapGroupMute);
  },

  async muteUser(roomId: string, input: { userId: string; durationHours: number; reason?: string }): Promise<GroupMute> {
    const response = await requestJson<{ mute: Record<string, unknown> }>(`/rooms/${roomId}/mutes`, {
      method: 'POST',
      body: JSON.stringify({
        user_id: input.userId,
        duration_hours: input.durationHours,
        ...(input.reason?.trim() ? { reason: input.reason.trim() } : {}),
      }),
    }, requireToken());
    return mapGroupMute(response.mute);
  },

  async unmuteUser(roomId: string, userId: string): Promise<void> {
    await requestJson(`/rooms/${roomId}/mutes/${userId}`, { method: 'DELETE' }, requireToken());
  },

  async updateGlobalMute(
    roomId: string,
    input: { enabled: boolean; reason?: string; durationMinutes?: number },
  ): Promise<GroupSettingsInfo> {
    const response = await requestJson<{ settings: Record<string, unknown> }>(`/rooms/${roomId}/mutes/global`, {
      method: 'POST',
      body: JSON.stringify({
        enabled: input.enabled,
        ...(input.reason?.trim() ? { reason: input.reason.trim() } : {}),
        ...(input.durationMinutes ? { duration_minutes: input.durationMinutes } : {}),
      }),
    }, requireToken());
    return mapGroupSettings(response.settings);
  },

  async listJoinRequests(roomId: string): Promise<GroupJoinRequest[]> {
    const response = await requestJson<{ requests?: Record<string, unknown>[] }>(
      `/rooms/${roomId}/join-requests`,
      {},
      requireToken(),
    );
    return (response.requests ?? []).map(mapGroupJoinRequest);
  },

  async createJoinRequest(roomId: string, message?: string): Promise<GroupJoinRequest> {
    const response = await requestJson<{ request: Record<string, unknown> }>(`/rooms/${roomId}/join-requests`, {
      method: 'POST',
      body: JSON.stringify({ ...(message?.trim() ? { message: message.trim() } : {}) }),
    }, requireToken());
    return mapGroupJoinRequest(response.request);
  },

  async reviewJoinRequest(
    roomId: string,
    requestId: string,
    status: Exclude<JoinRequestStatus, 'pending'>,
    reviewMessage?: string,
  ): Promise<GroupJoinRequest> {
    const response = await requestJson<{ request: Record<string, unknown> }>(
      `/rooms/${roomId}/join-requests/${requestId}/review`,
      {
        method: 'POST',
        body: JSON.stringify({
          status,
          ...(reviewMessage?.trim() ? { review_message: reviewMessage.trim() } : {}),
        }),
      },
      requireToken(),
    );
    return mapGroupJoinRequest(response.request);
  },

  async joinRoom(roomId: string): Promise<void> {
    await requestJson(`/rooms/${roomId}/join`, { method: 'POST' }, requireToken());
  },

  async updateJoinApproval(roomId: string, required: boolean): Promise<GroupSettingsInfo> {
    const response = await requestJson<{ settings: Record<string, unknown> }>(`/rooms/${roomId}/settings`, {
      method: 'PATCH',
      body: JSON.stringify({ join_approval_required: required }),
    }, requireToken());
    return mapGroupSettings(response.settings);
  },

  async listReceivedInvitations(status: GroupInvitationStatus | 'all' = 'all'): Promise<GroupInvitation[]> {
    const response = await requestJson<{ invitations?: Record<string, unknown>[] }>(
      withQuery('/group-invitations', { status }),
      {},
      requireToken(),
    );
    return (response.invitations ?? []).map(mapGroupInvitation);
  },

  async createInvitations(roomId: string, userIds: string[], message?: string): Promise<GroupInvitation[]> {
    const response = await requestJson<{ invitations?: Record<string, unknown>[] }>(`/rooms/${roomId}/invitations`, {
      method: 'POST',
      body: JSON.stringify({
        user_ids: userIds,
        ...(message?.trim() ? { message: message.trim() } : {}),
      }),
    }, requireToken());
    return (response.invitations ?? []).map(mapGroupInvitation);
  },

  async respondToInvitation(
    roomId: string,
    invitationId: string,
    status: Extract<GroupInvitationStatus, 'accepted' | 'declined'>,
  ): Promise<void> {
    await requestJson(`/rooms/${roomId}/invitations/${invitationId}/respond`, {
      method: 'PATCH',
      body: JSON.stringify({ status }),
    }, requireToken());
  },

  async fetchGroupSettings(roomId: string): Promise<GroupSettingsInfo> {
    const response = await requestJson<{ settings?: Record<string, unknown> } & Record<string, unknown>>(
      `/rooms/${roomId}/settings`,
      {},
      requireToken(),
    );
    return mapGroupSettings(response.settings ?? response);
  },

  async updateNotificationSettings(roomId: string, notificationSettings: number): Promise<void> {
    await requestJson(`/rooms/${roomId}/notification-settings`, {
      method: 'POST',
      body: JSON.stringify({ notification_settings: notificationSettings }),
    }, requireToken());
  },

  async pinRoom(roomId: string, pinned: boolean): Promise<void> {
    await requestJson(`/rooms/${roomId}/pin`, { method: pinned ? 'POST' : 'DELETE' }, requireToken());
  },

  async leaveRoom(roomId: string): Promise<void> {
    await requestJson(`/rooms/${roomId}/leave`, { method: 'POST' }, requireToken());
  },

  async dissolveRoom(roomId: string): Promise<void> {
    await requestJson(`/rooms/${roomId}`, { method: 'DELETE' }, requireToken());
  },

  async listOperationLogs(roomId: string, limit = 20, offset = 0): Promise<GroupOperationLog[]> {
    const response = await requestJson<{ logs?: Record<string, unknown>[] }>(
      withQuery(`/rooms/${roomId}/operation-logs`, { limit, offset }),
      {},
      requireToken(),
    );
    return (response.logs ?? []).map(mapGroupOperationLog);
  },
};
